// nats_service.proto - NATS 消息流服务 gRPC 接口
// 提供统一的事件总线和消息流接口，支持多租户和主题隔离
syntax = "proto3";

package isa.nats;

option go_package = "github.com/isa-cloud/isa_cloud/pkg/proto/nats";

import "common.proto";
import "google/protobuf/timestamp.proto";
import "google/protobuf/duration.proto";

// ========================================
// NATS 消息流服务
// ========================================
// 这是一个统一的消息流服务，为所有后端服务提供事件总线能力
// 支持：
// - 用户和组织级别的主题隔离
// - Pub/Sub 消息
// - 请求/响应模式
// - JetStream 持久化
// - 流管理
// - 消费者管理
// - 队列组

service NATSService {
    // ========== 基础 Pub/Sub ==========
    rpc Publish(PublishRequest) returns (PublishResponse);
    rpc PublishBatch(PublishBatchRequest) returns (PublishBatchResponse);
    rpc Subscribe(SubscribeRequest) returns (stream MessageResponse);
    rpc Unsubscribe(UnsubscribeRequest) returns (UnsubscribeResponse);
    
    // ========== 请求/响应模式 ==========
    rpc Request(RequestRequest) returns (RequestResponse);
    
    // ========== 队列组订阅 ==========
    rpc QueueSubscribe(QueueSubscribeRequest) returns (stream MessageResponse);
    
    // ========== JetStream 流管理 ==========
    rpc CreateStream(CreateStreamRequest) returns (CreateStreamResponse);
    rpc DeleteStream(DeleteStreamRequest) returns (DeleteStreamResponse);
    rpc GetStreamInfo(GetStreamInfoRequest) returns (GetStreamInfoResponse);
    rpc ListStreams(ListStreamsRequest) returns (ListStreamsResponse);
    rpc UpdateStream(UpdateStreamRequest) returns (UpdateStreamResponse);
    rpc PurgeStream(PurgeStreamRequest) returns (PurgeStreamResponse);
    
    // ========== JetStream 消费者管理 ==========
    rpc CreateConsumer(CreateConsumerRequest) returns (CreateConsumerResponse);
    rpc DeleteConsumer(DeleteConsumerRequest) returns (DeleteConsumerResponse);
    rpc GetConsumerInfo(GetConsumerInfoRequest) returns (GetConsumerInfoResponse);
    rpc ListConsumers(ListConsumersRequest) returns (ListConsumersResponse);
    
    // ========== JetStream 消息操作 ==========
    rpc PublishToStream(PublishToStreamRequest) returns (PublishToStreamResponse);
    rpc PullMessages(PullMessagesRequest) returns (PullMessagesResponse);
    rpc AckMessage(AckMessageRequest) returns (AckMessageResponse);
    rpc NakMessage(NakMessageRequest) returns (NakMessageResponse);
    
    // ========== 键值存储 (KV Store) ==========
    rpc KVPut(KVPutRequest) returns (KVPutResponse);
    rpc KVGet(KVGetRequest) returns (KVGetResponse);
    rpc KVDelete(KVDeleteRequest) returns (KVDeleteResponse);
    rpc KVKeys(KVKeysRequest) returns (KVKeysResponse);
    
    // ========== 对象存储 (Object Store) ==========
    rpc ObjectPut(ObjectPutRequest) returns (ObjectPutResponse);
    rpc ObjectGet(ObjectGetRequest) returns (ObjectGetResponse);
    rpc ObjectDelete(ObjectDeleteRequest) returns (ObjectDeleteResponse);
    rpc ObjectList(ObjectListRequest) returns (ObjectListResponse);
    
    // ========== 统计和监控 ==========
    rpc GetStatistics(GetStatisticsRequest) returns (GetStatisticsResponse);
    rpc GetStreamStats(GetStreamStatsRequest) returns (GetStreamStatsResponse);
    
    // ========== 健康检查 ==========
    rpc HealthCheck(NATSHealthCheckRequest) returns (NATSHealthCheckResponse);
}

// ========================================
// 数据结构
// ========================================

// 消息传递策略
enum DeliveryPolicy {
    DELIVERY_ALL = 0;                          // 所有消息
    DELIVERY_LAST = 1;                         // 最后一条
    DELIVERY_NEW = 2;                          // 新消息
    DELIVERY_BY_START_SEQUENCE = 3;            // 从序列号开始
    DELIVERY_BY_START_TIME = 4;                // 从时间开始
    DELIVERY_LAST_PER_SUBJECT = 5;             // 每个主题的最后一条
}

// 确认策略
enum AckPolicy {
    ACK_EXPLICIT = 0;                          // 显式确认
    ACK_ALL = 1;                               // 确认所有
    ACK_NONE = 2;                              // 不确认
}

// 重播策略
enum ReplayPolicy {
    REPLAY_INSTANT = 0;                        // 立即重播
    REPLAY_ORIGINAL = 1;                       // 按原始速度
}

// 存储类型
enum StorageType {
    STORAGE_FILE = 0;                          // 文件存储
    STORAGE_MEMORY = 1;                        // 内存存储
}

// 消息信息
message NATSMessage {
    string subject = 1;                        // 主题
    bytes data = 2;                            // 数据
    map<string, string> headers = 3;           // 消息头
    string reply_to = 4;                       // 回复主题
    google.protobuf.Timestamp timestamp = 5;   // 时间戳
}

// 流配置
message StreamConfig {
    string name = 1;                           // 流名称
    repeated string subjects = 2;              // 主题列表
    StorageType storage = 3;                   // 存储类型
    int32 max_msgs = 4;                        // 最大消息数
    int64 max_bytes = 5;                       // 最大字节数
    google.protobuf.Duration max_age = 6;      // 最大保留时间
    int32 max_msg_size = 7;                    // 最大消息大小
    int32 replicas = 8;                        // 副本数
    bool discard_new_per_subject = 9;          // 每个主题丢弃新消息
}

// 流信息
message StreamInfo {
    string name = 1;
    StreamConfig config = 2;
    StreamState state = 3;
    google.protobuf.Timestamp created = 4;
}

// 流状态
message StreamState {
    int64 messages = 1;                        // 消息总数
    int64 bytes = 2;                           // 字节总数
    int64 first_seq = 3;                       // 第一条消息序列号
    int64 last_seq = 4;                        // 最后一条消息序列号
    google.protobuf.Timestamp first_ts = 5;    // 第一条消息时间
    google.protobuf.Timestamp last_ts = 6;     // 最后一条消息时间
    int32 num_subjects = 7;                    // 主题数量
    int32 consumer_count = 8;                  // 消费者数量
}

// 消费者配置
message ConsumerConfig {
    string name = 1;                           // 消费者名称
    string durable_name = 2;                   // 持久化名称
    string filter_subject = 3;                 // 过滤主题
    DeliveryPolicy delivery_policy = 4;        // 传递策略
    AckPolicy ack_policy = 5;                  // 确认策略
    google.protobuf.Duration ack_wait = 6;     // 确认等待时间
    int32 max_deliver = 7;                     // 最大传递次数
    ReplayPolicy replay_policy = 8;            // 重播策略
    int64 opt_start_seq = 9;                   // 起始序列号
    google.protobuf.Timestamp opt_start_time = 10; // 起始时间
}

// 消费者信息
message ConsumerInfo {
    string stream_name = 1;
    string name = 2;
    ConsumerConfig config = 3;
    ConsumerState state = 4;
    google.protobuf.Timestamp created = 5;
}

// 消费者状态
message ConsumerState {
    int64 delivered_seq = 1;                   // 已传递序列号
    int64 ack_floor_seq = 2;                   // 已确认序列号
    int64 num_pending = 3;                     // 待处理数量
    int64 num_redelivered = 4;                 // 重新传递数量
}

// ========================================
// 基础 Pub/Sub 请求/响应
// ========================================

message PublishRequest {
    string user_id = 1;
    string organization_id = 2;
    string subject = 3;
    bytes data = 4;
    map<string, string> headers = 5;
    string reply_to = 6;
}

message PublishResponse {
    bool success = 1;
    string message = 2;
}

message PublishBatchRequest {
    string user_id = 1;
    string organization_id = 2;
    repeated NATSMessage messages = 3;
}

message PublishBatchResponse {
    bool success = 1;
    int32 published_count = 2;
    repeated string errors = 3;
}

message SubscribeRequest {
    string user_id = 1;
    string organization_id = 2;
    string subject = 3;                        // 支持通配符 (*, >)
    string queue_group = 4;                    // 可选队列组
}

message MessageResponse {
    string subject = 1;
    bytes data = 2;
    map<string, string> headers = 3;
    string reply_to = 4;
    google.protobuf.Timestamp timestamp = 5;
    int64 sequence = 6;                        // JetStream 序列号（如果有）
}

message UnsubscribeRequest {
    string user_id = 1;
    string organization_id = 2;
    string subject = 3;
}

message UnsubscribeResponse {
    bool success = 1;
}

// ========================================
// 请求/响应模式
// ========================================

message RequestRequest {
    string user_id = 1;
    string organization_id = 2;
    string subject = 3;
    bytes data = 4;
    google.protobuf.Duration timeout = 5;
}

message RequestResponse {
    bool success = 1;
    bytes data = 2;
    string error_message = 3;
}

// ========================================
// 队列组订阅
// ========================================

message QueueSubscribeRequest {
    string user_id = 1;
    string organization_id = 2;
    string subject = 3;
    string queue_group = 4;
}

// ========================================
// JetStream 流管理
// ========================================

message CreateStreamRequest {
    string user_id = 1;
    string organization_id = 2;
    StreamConfig config = 3;
}

message CreateStreamResponse {
    bool success = 1;
    StreamInfo stream = 2;
    string message = 3;
}

message DeleteStreamRequest {
    string user_id = 1;
    string organization_id = 2;
    string stream_name = 3;
}

message DeleteStreamResponse {
    bool success = 1;
}

message GetStreamInfoRequest {
    string user_id = 1;
    string organization_id = 2;
    string stream_name = 3;
}

message GetStreamInfoResponse {
    StreamInfo stream = 1;
}

message ListStreamsRequest {
    string user_id = 1;
    string organization_id = 2;
}

message ListStreamsResponse {
    repeated StreamInfo streams = 1;
}

message UpdateStreamRequest {
    string user_id = 1;
    string organization_id = 2;
    string stream_name = 3;
    StreamConfig config = 4;
}

message UpdateStreamResponse {
    bool success = 1;
    StreamInfo stream = 2;
}

message PurgeStreamRequest {
    string user_id = 1;
    string organization_id = 2;
    string stream_name = 3;
    string subject = 4;                        // 可选，清除特定主题
}

message PurgeStreamResponse {
    bool success = 1;
    int64 purged_count = 2;
}

// ========================================
// JetStream 消费者管理
// ========================================

message CreateConsumerRequest {
    string user_id = 1;
    string organization_id = 2;
    string stream_name = 3;
    ConsumerConfig config = 4;
}

message CreateConsumerResponse {
    bool success = 1;
    ConsumerInfo consumer = 2;
    string message = 3;
}

message DeleteConsumerRequest {
    string user_id = 1;
    string organization_id = 2;
    string stream_name = 3;
    string consumer_name = 4;
}

message DeleteConsumerResponse {
    bool success = 1;
}

message GetConsumerInfoRequest {
    string user_id = 1;
    string organization_id = 2;
    string stream_name = 3;
    string consumer_name = 4;
}

message GetConsumerInfoResponse {
    ConsumerInfo consumer = 1;
}

message ListConsumersRequest {
    string user_id = 1;
    string organization_id = 2;
    string stream_name = 3;
}

message ListConsumersResponse {
    repeated ConsumerInfo consumers = 1;
}

// ========================================
// JetStream 消息操作
// ========================================

message PublishToStreamRequest {
    string user_id = 1;
    string organization_id = 2;
    string stream_name = 3;
    string subject = 4;
    bytes data = 5;
    map<string, string> headers = 6;
}

message PublishToStreamResponse {
    bool success = 1;
    int64 sequence = 2;                        // 消息序列号
    google.protobuf.Timestamp timestamp = 3;
    string message = 4;
}

message PullMessagesRequest {
    string user_id = 1;
    string organization_id = 2;
    string stream_name = 3;
    string consumer_name = 4;
    int32 batch_size = 5;                      // 拉取数量
    google.protobuf.Duration max_wait = 6;     // 最大等待时间
}

message PullMessagesResponse {
    repeated JetStreamMessage messages = 1;
}

message JetStreamMessage {
    string subject = 1;
    bytes data = 2;
    map<string, string> headers = 3;
    int64 sequence = 4;
    google.protobuf.Timestamp timestamp = 5;
    int32 num_delivered = 6;                   // 传递次数
}

message AckMessageRequest {
    string user_id = 1;
    string organization_id = 2;
    string stream_name = 3;
    string consumer_name = 4;
    int64 sequence = 5;
}

message AckMessageResponse {
    bool success = 1;
}

message NakMessageRequest {
    string user_id = 1;
    string organization_id = 2;
    string stream_name = 3;
    string consumer_name = 4;
    int64 sequence = 5;
    google.protobuf.Duration delay = 6;        // 延迟重新传递
}

message NakMessageResponse {
    bool success = 1;
}

// ========================================
// 键值存储请求/响应
// ========================================

message KVPutRequest {
    string user_id = 1;
    string organization_id = 2;
    string bucket = 3;
    string key = 4;
    bytes value = 5;
}

message KVPutResponse {
    bool success = 1;
    int64 revision = 2;
}

message KVGetRequest {
    string user_id = 1;
    string organization_id = 2;
    string bucket = 3;
    string key = 4;
}

message KVGetResponse {
    bool found = 1;
    bytes value = 2;
    int64 revision = 3;
    google.protobuf.Timestamp created = 4;
}

message KVDeleteRequest {
    string user_id = 1;
    string organization_id = 2;
    string bucket = 3;
    string key = 4;
}

message KVDeleteResponse {
    bool success = 1;
}

message KVKeysRequest {
    string user_id = 1;
    string organization_id = 2;
    string bucket = 3;
}

message KVKeysResponse {
    repeated string keys = 1;
}

// ========================================
// 对象存储请求/响应
// ========================================

message ObjectPutRequest {
    string user_id = 1;
    string organization_id = 2;
    string bucket = 3;
    string object_name = 4;
    bytes data = 5;
    map<string, string> metadata = 6;
}

message ObjectPutResponse {
    bool success = 1;
    string object_id = 2;
}

message ObjectGetRequest {
    string user_id = 1;
    string organization_id = 2;
    string bucket = 3;
    string object_name = 4;
}

message ObjectGetResponse {
    bool found = 1;
    bytes data = 2;
    map<string, string> metadata = 3;
}

message ObjectDeleteRequest {
    string user_id = 1;
    string organization_id = 2;
    string bucket = 3;
    string object_name = 4;
}

message ObjectDeleteResponse {
    bool success = 1;
}

message ObjectListRequest {
    string user_id = 1;
    string organization_id = 2;
    string bucket = 3;
}

message ObjectListResponse {
    repeated ObjectInfo objects = 1;
}

message ObjectInfo {
    string name = 1;
    int64 size = 2;
    google.protobuf.Timestamp modified = 3;
    map<string, string> metadata = 4;
}

// ========================================
// 统计和监控请求/响应
// ========================================

message GetStatisticsRequest {
    string user_id = 1;
    string organization_id = 2;
}

message GetStatisticsResponse {
    int64 total_streams = 1;
    int64 total_consumers = 2;
    int64 total_messages = 3;
    int64 total_bytes = 4;
    int32 connections = 5;
    int64 in_msgs = 6;
    int64 out_msgs = 7;
    int64 in_bytes = 8;
    int64 out_bytes = 9;
}

message GetStreamStatsRequest {
    string user_id = 1;
    string organization_id = 2;
    string stream_name = 3;
}

message GetStreamStatsResponse {
    string stream_name = 1;
    int64 messages = 2;
    int64 bytes = 3;
    int32 consumer_count = 4;
    double message_rate = 5;                   // 消息速率（msg/s）
    double byte_rate = 6;                      // 字节速率（bytes/s）
}

// ========================================
// 健康检查
// ========================================

message NATSHealthCheckRequest {
    bool deep_check = 1;
}

message NATSHealthCheckResponse {
    bool healthy = 1;
    string nats_status = 2;
    bool jetstream_enabled = 3;
    int32 connections = 4;
    google.protobuf.Timestamp checked_at = 5;
    string message = 6;
}



